### PR TITLE
Single threaded tracer

### PR DESCRIPTION
**Pull request by rnburn**
The goal of this PR is to provide a tracer that does no internal threading and allows a customizable transport while not exposing internal data structures used by the library.

I put together an [example](https://github.com/rnburn/lightstep-tracer-cpp-1/blob/single-threaded-tracer/c%2B%2B11/example/manual_tracer/qa_server.cpp) that shows how the tracer could be integrated into an application using libevent.

There are still quite a few pieces to fill in for the library. Some of the design points I've been thinking about are
* How should you opt in to use the single threaded tracer? Currently I have it set up to use a separate [options structure](https://github.com/rnburn/lightstep-tracer-cpp-1/blob/single-threaded-tracer/c%2B%2B11/include/lightstep/tracer.h#L52), but not sure that's the best way. 
* Do we want to provide a default transporter for the single threaded tracer? I wrote out what [the API](https://github.com/rnburn/lightstep-tracer-cpp-1/blob/single-threaded-tracer/c%2B%2B11/src/nghttp2_async_transporter.h#L8) might look like and showed how you could use it in the [example](https://github.com/rnburn/lightstep-tracer-cpp-1/blob/single-threaded-tracer/c%2B%2B11/example/manual_tracer/qa_server.cpp#L40), but didn't provide an implementation. GRPC's [async api](https://grpc.io/docs/tutorials/async/helloasync-cpp.html) doesn't let you work at the socket level, so you would need to use a lower level library like [nghttp2](https://nghttp2.org/) if you wanted to make a default transporter that's usable in applications built on top of [epoll](https://en.wikipedia.org/wiki/Epoll) or equivalent. This version of the tracer seems pretty specialized, so I'm not sure who would use such a default transporter.
* I image you'd only want to be recording spans on a single thread for this tracer, but do we want to still make the recorder thread-safe in case it is called concurrently for some reason?

### CODE REVIEW

**Comment by jmacd** [comment]
This looks good. To the questions (1) and (3) above re: options struct and synchronization, one solution I would favor is to force the user to subclass their instance, then (1) make abstract virtual getters for the things you need and the user provide those options in code, (3) let the user provide locking if they need it for some reason. Unless this solution is much worse, that is.

For (2) I think you've done enough, and in any case I would be most interested in what Lyft needs to upgrade to this version immediately for Envoy. As we have customers that want to use this arrive, we may know more about examples that would be useful to have.

This is definitely on a good track.

File: c++11/src/grpc_transporter.cpp [hunk 1]
```diff
@@ -13,34 +19,64 @@ static std::string HostPortOf(const LightStepTracerOptions& options) {
...
+      const google::protobuf::Message& request,
+      google::protobuf::Message& response) override {
+    grpc::ClientContext context;
+    collector::ReportResponse resp;
+    context.set_fail_fast(true);
+    context.set_deadline(std::chrono::system_clock::now() + report_timeout_);
+    auto status = client_.Report(
+        &context, dynamic_cast<const collector::ReportRequest&>(request),
+        dynamic_cast<collector::ReportResponse*>(&response));
+    if (!status.ok()) {
```
Comment (jmacd) [inline c++11/src/grpc_transporter.cpp:70]: the `status` object has an associated code number. Is there a way to use that?

**Review by jmacd — APPROVED** [review]

File: c++11/src/tracer.cpp [hunk 2]
```diff
@@ -66,10 +86,60 @@ LightStepTracer::MakeSpanContext(
...
+    transporter = MakeGrpcTransporter(*logger, options);
+  }
+  auto recorder = std::unique_ptr<Recorder>{
+      new AutoRecorder{*logger, std::move(options), std::move(transporter)}};
+  return std::shared_ptr<LightStepTracer>{
+      new LightStepTracerImpl{std::move(logger), std::move(recorder)}};
+}
+
+//------------------------------------------------------------------------------
+// MakeNonThreadedTracer
```
Comment (jmacd) [inline c++11/src/tracer.cpp]: nit: I'd prefer to call this MakeSingleThreadedTracer. any reason that's less clear?

Comment (rnburn) [reply c++11/src/tracer.cpp]: That's better. I'll change it.

File: c++11/src/grpc_transporter.cpp (hunk 1 above)
Comment (rnburn) [reply c++11/src/grpc_transporter.cpp:70]: I could probably use it to map to better standard [error codes](http://www.cplusplus.com/reference/system_error/errc/). Using them directly would mean having to set up a gRPC specific [error category](http://en.cppreference.com/w/cpp/error/error_category).

**Pull request closed without merging**